Mathlib/Tactic/GRW/Core.lean

```diff
@@ -0,0 +1,179 @@
+/-
+Copyright (c) 2023 Sebastian Zimmer. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Sebastian Zimmer, Mario Carneiro, Heather Macbeth
+-/
+import Mathlib.Tactic.Core
+import Std.Tactic.LabelAttr
+import Mathlib.Tactic.GCongr.Core
+import Mathlib.Data.Prod.Basic
+
+/-!
+# GRW Tactic
+
+This module defines the core of the `grw` tactic.
+
+-/
+
+open Lean Meta Std.Tactic.LabelAttr Elab Tactic
+
+namespace Mathlib.Tactic.GRW
+
+initialize registerTraceClass `GRW
+
+/--
+Lemmas marked `@[grw]` are used by the `grw` tactic to use a relation to rewrite an expression.
+
+The lemma is used to rewrite its first explicit argument into the result type. The other arguments
+are then filled in using the `gcongr` tactic.
+
+For example, this lemma shows `grw` how to convert `a < b` into `c < d`.
+
+```lean
+@[grw]
+lemma rewrite_lt {α : Type} [Preorder α] {a b c d : α} (h₁ : a < b) (h₂ : c ≤ a) (h₃ : b ≤ d) :
+    c < d := lt_of_le_of_lt h₂ (lt_of_lt_of_le h₁ h₃)
+```
+
+These lemmas can do more than just use transitivity. This lemma shows `grw` to to rewrite `a ∈ X`
+into `a ∈ Y`.
+
+```lean
+@[grw]
+lemma rewrite_mem {α : Type} {a : α} {X Y: Set α} (h₁ : a ∈ X) (h₂ : X ⊆ Y) : a ∈ Y := h₂ h₁
+```
+-/
+initialize ext : LabelExtension ← (
+  let descr := "A lemma describing how to convert the first argument into the target type, possibly
+introducing side goals. These side goals will be solved with `gcongr`"
+  let grw := `grw
+  registerLabelAttr grw descr grw)
+
+open GCongr
+
+/-- Given `relationType := R a b`, returns `(a, b)` -/
```

Comment:
Should this be counting explicit arguments? I'm considering a relation which has some trailing typeclass arguments (as an extreme case, `MonoidHom`!)